usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +519,519 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
     return CKR_OK;
 }
 
+/*******************************************************************************
+ *
+ *                    Begin EP11 protected key option
+ */
+
+typedef struct {
+    CK_BBOOL wrap_was_successful;
+    CK_VOID_PTR secure_key;
+    CK_ULONG secure_key_len;
+    CK_BYTE *pkey_buf;
+    size_t *pkey_buflen_p;
+} pkey_wrap_handler_data_t;
+
+/* A wrapped key can have max 132 bytes for an EC-p521 key */
+#define EP11_MAX_WRAPPED_KEY_SIZE      (2 * (521 / 8 + 1))
+#define EP11_WRAPPED_KEY_VERSION_1     0x0001
+#define EP11_WRAPPED_KEY_TYPE_AES      0x1
+#define EP11_WRAPPED_KEY_TYPE_DES      0x2
+#define EP11_WRAPPED_KEY_TYPE_EC       0x3
+#define EP11_WRAPPED_KEY_TYPE_ED       0x4
+typedef struct {
+    uint16_t version;
+    uint8_t res0[16];
+    uint32_t wrapped_key_type;
+    uint32_t bit_length;
+    uint64_t token_size;
+    uint8_t res1[8];
+    uint8_t wrapped_key[EP11_MAX_WRAPPED_KEY_SIZE];
+    uint8_t res2[50];
+} __attribute__((packed)) wrapped_key_t;
+
+/**
+ * Callback function used by handle_all_ep11_cards() for creating a protected
+ * key via the given APQN (adaper,domain).
+ * Note that this function only works with an ep11 host lib v3 or later,
+ * because since v3 the target is a numeric value and we can OR the
+ * XCP_TGTFL_SET_SCMD flag with it. Before calling this function, it has been
+ * checked if running with v3 or later and the CPACF_WRAP mechanism is
+ * supported by the hw.
+ */
+static CK_RV ep11tok_pkey_wrap_handler(uint_32 adapter, uint_32 domain,
+                                       void *handler_data)
+{
+    CK_BYTE iv[16] = {
+        0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
+        0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10
+    };
+    CK_MECHANISM mech = { CKM_IBM_CPACF_WRAP, &iv, sizeof(iv) };
+    pkey_wrap_handler_data_t *data = (pkey_wrap_handler_data_t *) handler_data;
+    target_t target = 0;
+    CK_RV ret;
+
+    if (data->wrap_was_successful)
+        goto done;
+
+    ret = get_ep11_target_for_apqn(adapter, domain, &target, XCP_MFL_PROBE);
+    if (ret != CKR_OK)
+        goto done;
+
+    /* Create the protected key via CKM_IBM_CPACF_WRAP */
+    ret = dll_m_WrapKey(data->secure_key, data->secure_key_len,
```

Comment:
I guess here is something missing. The way it currently is will cause failures when STRUCT_SESSION mode or VHSM_MODE is enabled for the EP11 token. Will do some further tests and let you know offline what changes are required..

Reply:
Well, no, all is OK. dll_m_WrapKey does not have/need a pin blob, so no need to do anything special for STRICT_SESSION mode or VHSM_MODE.

Please ignore above comment ;-)